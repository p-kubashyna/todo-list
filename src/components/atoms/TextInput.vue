<script setup>
import { ref } from "vue";
import store from "@/store";

const inputText = ref("");
const isInvalidInput = ref(false);

function submitTodo(inputText) {
  this.isInvalidInput = validateInput(this.inputText);
  if (!this.isInvalidInput) {
    store.commit("addTodo", inputText);
    this.inputText = "";
  }
}

function validateInput(inputText) {
  return !inputText || /^\s*$/.test(inputText) || inputText.length > 150;
}
</script>

<template>
  <form @submit.prevent="submitTodo(inputText)">
    <input
      class="input-base"
      :class="{ 'input-invalid': isInvalidInput === true }"
      v-model.trim="inputText"
      placeholder="Add task..."
    />
  </form>
</template>

<style scoped>
.input-base {
  width: 600px;
  height: 25px;
  border-radius: 7px;
  border-width: 10px;
  border-color: white;
  border-style: solid;
  font-family: monospace;
}

.input-invalid {
  background: white url(../../assets/images/invalid_input_icon.png) no-repeat
    right;
  background-size: auto 100%;
}
</style>
