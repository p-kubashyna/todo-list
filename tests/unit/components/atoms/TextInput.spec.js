import { describe, expect, it } from "@jest/globals";
import { shallowMount } from "@vue/test-utils";
import TextInput from "@/components/atoms/TextInput.vue";

describe("Given the TextInput component", () => {
  const defaultPros = { placeholder: "placeholder", isInvalidInput: true };

  it.each([
    [{ placeholder: "placeholder1", isInvalidInput: true }],
    [{ placeholder: "placeholder2", isInvalidInput: false }],
  ])("should render", (propsData) => {
    const wrapper = shallowMount(TextInput, { props: propsData });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should set the value", async () => {
    const wrapper = shallowMount(TextInput, { props: defaultPros });
    const textInput = wrapper.find("input");
    await textInput.setValue("some value");

    expect(wrapper.find("input").element.value).toBe("some value");
    expect(wrapper.vm.inputText).toBe("some value");
  });

  it("should emit an event and clear input when form is submitted", async () => {
    const wrapper = shallowMount(TextInput, { props: defaultPros });
    await wrapper.find("form").trigger("submit");
    const submitFormEvent = wrapper.emitted("submitForm");

    expect(wrapper.emitted()).toHaveProperty("submitForm");
    expect(submitFormEvent).toHaveLength(1);
    expect(wrapper.vm.inputText).toBe("");
  });
});
