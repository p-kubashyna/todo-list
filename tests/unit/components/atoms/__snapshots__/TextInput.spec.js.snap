// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the TextInput component should render 1`] = `
<form>
  <input
    class="input-base input-invalid"
    placeholder="placeholder1"
  />
</form>
`;

exports[`Given the TextInput component should render 2`] = `
<form>
  <input
    class="input-base"
    placeholder="placeholder2"
  />
</form>
`;
